carling/categorical.py

```diff
@@ -137,22 +137,32 @@ def CreateCategoricalDicts(pcoll, cat_cols, existing_dict_rows):
 
 
 @beam.ptransform_fn
-def ReplaceCategoricalColumns(inputs, categorical_dict_rows):
+def ReplaceCategoricalColumns(
+    inputs, cat_cols, categorical_dict_rows, default_unseen=None
+):
     """
     Utilizes the "categorical dictionary rows" generated by
     CreateCategoricalDicts which is a list of pairs of type of (column, value,
     unique_int), and replaces each column with the appropriate value found in
     the mapping.
+
+    In there case where no mapping exists, `default_unseen` will be returned.
+    This defaults to None, but can be set as a kwarg.
     """
 
-    categorical_mapping_dicts = categorical_dict_rows | beam.Map(
-        lambda r: ((r[0], r[1]), r[2])
+    categorical_mapping_dicts = (
+        categorical_dict_rows
+        | "index by column" >> beam.Map(lambda r: (r[0], (r[1], r[2])))
+        | "group by column" >> beam.GroupByKey()
+        | "create dict per column" >> beam.Map(lambda r: (r[0], dict(r[1])))
     )
 
+    def _process(row, d):
+        for col in cat_cols:
+            row[col] = d.get(col, {}).get(row[col], default_unseen)
```

Comment:
According to the Apache Beam's document, element should not be mutated.
It is better to create a new dictionary instead.

reference: <https://beam.apache.org/releases/pydoc/2.27.0/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.process>

Reply:
974418e